# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

In [3]:
data_path = 'data/supermarket_sales.csv'
df = pd.read_csv(data_path)

In [4]:
df

Invoice ID Branch       City Customer type  Gender  \
0    750-67-8428      A     Yangon        Member  Female   
1    226-31-3081      C  Naypyitaw        Normal  Female   
2    631-41-3108      A     Yangon        Normal    Male   
3    123-19-1176      A     Yangon        Member    Male   
4    373-73-7910      A     Yangon        Normal    Male   
..           ...    ...        ...           ...     ...   
995  233-67-5758      C  Naypyitaw        Normal    Male   
996  303-96-2227      B   Mandalay        Normal  Female   
997  727-02-1313      A     Yangon        Member    Male   
998  347-56-2442      A     Yangon        Normal    Male   
999  849-09-3807      A     Yangon        Member  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
0         Health and beauty       74.69         7  26.1415   548.9715   
1    Electronic accessories       15.28         5   3.8200    80.2200   
2        Home and lifestyle       46.33         7  16.2155   340.5255   
3         Health and beauty       58.22         8  23.2880   489.0480   
4         Sports and travel       86.31         7  30.2085   634.3785   
..                      ...         ...       ...      ...        ...   
995       Health and beauty       40.35         1   2.0175    42.3675   
996      Home and lifestyle       97.38        10  48.6900  1022.4900   
997      Food and beverages       31.84         1   1.5920    33.4320   
998      Home and lifestyle       65.82         1   3.2910    69.1110   
999     Fashion accessories       88.34         7  30.9190   649.2990   

          Date   Time      Payment    cogs  gross margin percentage  \
0     1/5/2019  13:08      Ewallet  522.83                 4.761905   
1     3/8/2019  10:29         Cash   76.40                 4.761905   
2     3/3/2019  13:23  Credit card  324.31                 4.761905   
3    1/27/2019  20:33      Ewallet  465.76                 4.761905   
4     2/8/2019  10:37      Ewallet  604.17                 4.761905   
..         ...    ...          ...     ...                      ...   
995  1/29/2019  13:46      Ewallet   40.35                 4.761905   
996   3/2/2019  17:16      Ewallet  973.80                 4.761905   
997   2/9/2019  13:22         Cash   31.84                 4.761905   
998  2/22/2019  15:33         Cash   65.82                 4.761905   
999  2/18/2019  13:28         Cash  618.38                 4.761905   

     gross income  Rating  
0         26.1415     9.1  
1          3.8200     9.6  
2         16.2155     7.4  
3         23.2880     8.4  
4         30.2085     5.3  
..            ...     ...  
995        2.0175     6.2  
996       48.6900     4.4  
997        1.5920     7.7  
998        3.2910     4.1  
999       30.9190     6.6  

[1000 rows x 17 columns]

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [5]:
df.sort_values('Unit price', ascending=False).head(5)

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [6]:
df[df['Product line'] == 'Electronic accessories'].nlargest(1,['Unit price','Quantity'])

Invoice ID Branch      City Customer type  Gender  \
209  817-69-8206      B  Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
209  Electronic accessories       99.73         9  44.8785  942.4485   

         Date   Time      Payment    cogs  gross margin percentage  \
209  3/2/2019  19:42  Credit card  897.57                 4.761905   

     gross income  Rating  
209       44.8785     6.5

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [7]:
# Method 2

# Step 1. Create a new column 'Rating Category' using conditions
def category_rating(val):
    if val >= 8:
        return 'High'
    elif val >= 6 and val < 8:
        return 'Medium'
    else:
        return 'Low'
        
# Step 2. Apply the function to create the new column
df['Rating Category'] = df.Rating.apply(category_rating)

# Step 3. Display the first ten rows to verify the new column
df[['Rating Category','Rating']].head(10)

Rating Category  Rating
0            High     9.1
1            High     9.6
2          Medium     7.4
3            High     8.4
4             Low     5.3
5             Low     4.1
6             Low     5.8
7            High     8.0
8          Medium     7.2
9             Low     5.9

In [8]:
# Method 2

# Step 1. Create a new column 'Rating Category' using conditions
df['Rating Category'] = pd.cut(df['Rating'],
                              bins=[0, 6, 8, float('inf')],
                              labels=['Low', 'Medium', 'High'],
                              include_lowest=True)

# Step 2. Display the first ten rows to verify the new column
df[['Rating Category','Rating']].head(10)

Rating Category  Rating
0            High     9.1
1            High     9.6
2          Medium     7.4
3            High     8.4
4             Low     5.3
5             Low     4.1
6             Low     5.8
7          Medium     8.0
8          Medium     7.2
9             Low     5.9

In [9]:
# Method 3
import numpy as np

# Step 1. Define conditions and choices
conditions = [
    df['Rating'] >= 8,
    (df['Rating'] >= 6) & (df['Rating'] < 8),
    df['Rating'] < 6
]
choices = ['High', 'Medium', 'Low']

# Step 2. Create new column using np.select
df['Rating Category'] = np.select(conditions, choices)

# Step 3. Display the first ten rows to verify the new column
df[['Rating Category','Rating']].head(10)

Rating Category  Rating
0            High     9.1
1            High     9.6
2          Medium     7.4
3            High     8.4
4             Low     5.3
5             Low     4.1
6             Low     5.8
7            High     8.0
8          Medium     7.2
9             Low     5.9

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [10]:
df['Rating Category'].value_counts(normalize = True)*100

Rating Category
Medium    35.6
High      32.9
Low       31.5
Name: proportion, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [11]:
time = '11:28'
type(time)

str

In [12]:
def get_hour(time_str):
    return int(time_str.split(':')[0])

In [13]:
df['Hour'] = df.Time.apply(get_hour)
df[['Product line','Unit price','Hour']].head(10)

Product line  Unit price  Hour
0       Health and beauty       74.69    13
1  Electronic accessories       15.28    10
2      Home and lifestyle       46.33    13
3       Health and beauty       58.22    20
4       Sports and travel       86.31    10
5  Electronic accessories       85.39    18
6  Electronic accessories       68.84    14
7      Home and lifestyle       73.56    11
8       Health and beauty       36.26    17
9      Food and beverages       54.84    13

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [14]:
df['Hour'] = df['Time'].apply(get_hour)
df[['Time', 'Hour']].head(10)

Time  Hour
0  13:08    13
1  10:29    10
2  13:23    13
3  20:33    20
4  10:37    10
5  18:30    18
6  14:36    14
7  11:38    11
8  17:15    17
9  13:27    13

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [15]:
invoices_per_hour = df['Hour'].value_counts()

print("Number of Purchases by Hour:")
print(invoices_per_hour)

Number of Purchases by Hour:
Hour
19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: count, dtype: int64


In [16]:
invoices_per_hour_percentages = (df['Hour'].value_counts(normalize=True) * 100)

print("Percentage Distribution of Purchases by Hour:")
print(invoices_per_hour_percentages)

Percentage Distribution of Purchases by Hour:
Hour
19    11.3
13    10.3
15    10.2
10    10.1
18     9.3
11     9.0
12     8.9
14     8.3
16     7.7
20     7.5
17     7.4
Name: proportion, dtype: float64


**Analysis of Purchase Hours**

After analyzing the number of purchases by hour, we can see that the highest number of purchases occurs at 7 PM (113 = 11,3%) 1 PM (103 = 10,3%), 3 PM (102 = 10.2% ).

This seems logical:

- 1 PM – lunchtime, when many people have a break and can visit a store.
- 3 PM – the active part of the afternoon, possibly related to errands or shopping before the workday ends.
- 7 PM – after working hours, when people are heading home and may stop by a store.

So, the peak purchasing hours are during lunch and after work, which aligns with most people’s typical daily routines. This insight can be used to:

- better schedule staff shifts (increase staff during peak hours),
- plan promotions during off-peak times,
- optimize checkout operations.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [23]:
avg_total_per_city = df.groupby('City')['Total'].mean().round(2)

print("Average Receipt Value by City:")
print(avg_total_per_city)

Average Receipt Value by City:
City
Mandalay     319.87
Naypyitaw    337.10
Yangon       312.35
Name: Total, dtype: float64


6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [18]:
agg_Total = df.groupby('City')['Total'].agg(['count','mean','min','max'])
agg_Total

count        mean      min      max
City                                          
Mandalay     332  319.872506  18.6375  1022.49
Naypyitaw    328  337.099715  10.6785  1042.65
Yangon       340  312.354031  12.6945  1039.29

**Insights and Recommendations**
1. Highest average spending per transaction:
Naypyitaw has the highest average total (337.10), indicating that on average, customers there spend more per purchase compared to Mandalay (319.87) and Yangon (312.35).
2. Largest number of purchases:
Yangon has the highest number of purchases (340), suggesting greater foot traffic or a larger customer base, even though the average transaction value is the lowest among the three cities.
3. Price sensitivity or customer profile differences:
Naypyitaw also has the lowest minimum transaction value (10.68), which may suggest a more diverse customer base — from very small to large spenders.

**Recommendations for the supermarket chain:**
- For Naypyitaw:
This city seems to have high potential for premium products or bulk sales. Consider expanding product offerings in higher price segments or introducing loyalty programs for big spenders.
- For Yangon:
Focus on increasing the average check. This can be done through:
    - Product bundling or cross-selling strategies.
    - Upselling promotions at checkout.
    - Promotions targeting regular but small-spending customers.
- For Mandalay:
Performance is balanced. To grow further:
    - Study best practices from Naypyitaw and Yangon (e.g., product mix or promotions).
    - Analyze customer preferences more deeply to tailor campaigns.


7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [19]:
df.to_csv('data/supermarket_sales_enriched.csv', index=False)

In [21]:
df_check = pd.read_csv('data/supermarket_sales_enriched.csv')
df_check.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  Hour  
0            High    13  
1            High    10  
2          Medium    13  
3            High    20  
4             Low    10